File: Python/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head is None:
            return True
        p, q = head, head
        while q.next and q.next.next:
            p = p.next
            q = q.next.next
        p = p.next
        re = None
        while p:
            tmp = p.next
            p.next = re
            re = p
            p = tmp
        while re:
            if re.val != head.val:
                return False
            re = re.next
            head = head.next
        return True

File: C++/234. Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* p = head;
        ListNode* q = head;
        if(head == nullptr)
            return true;
        while(q->next && q->next->next)
        {
            p = p->next;
            q = q->next->next;
        }
        p = p->next;
        ListNode* re = nullptr;
        while(p)
        {
            ListNode* tmp = p->next;
            p->next = re;
            re = p;
            p = tmp;
        }
        while(re)
        {
            if(head->val != re->val)
                return false;
            head = head->next;
            re = re->next;
        }
        return true;
    }
};